mlir/lib/Dialect/Linalg/TransformOps/Syntax.cpp

```diff
@@ -12,9 +12,13 @@
 using namespace mlir;
 
 ParseResult mlir::parseSemiFunctionType(OpAsmParser &parser, Type &argumentType,
-                                        Type &resultType) {
+                                        Type &resultType, bool resultOptional) {
   argumentType = resultType = nullptr;
-  bool hasLParen = parser.parseOptionalLParen().succeeded();
+
+  bool hasLParen = resultOptional ? parser.parseOptionalLParen().succeeded()
+                                  : parser.parseLParen().succeeded();
+  if (!resultOptional && !hasLParen)
+    return failure();
   if (parser.parseType(argumentType).failed())
     return failure();
   if (!hasLParen)
```

Comment:
Does this need the new `resultOptional` option? Does it work if you do something like:
```cpp
  argumentType = resultType = nullptr;
  bool hasLParen = parser.parseOptionalLParen().succeeded();
  if (parser.parseType(argumentType).failed())
    return failure();
  if (!hasLParen)
    return success(parser.parseRParen().failed() ||
                   parser.parseArrow().failed() ||
                   parser.parseType(resultType).failed());
  return failure(parser.parseRParen().failed() ||
                 parser.parseArrow().failed() ||
                 parser.parseType(resultType).failed());
```

Reply:
This causes problems for Ops with exactly one non optional result, like `transform.structured.generalize`. We could write an invalid version of such an op like this:
```mlir
transform.structured.generalize %arg0 : !transform.any_op
```
This would successfully go through your parser because it looks like a valid `SemiFunctionType` with optional or no results. However, this will then lead to a crash during Op creation, because `resultType` is never initialized, yet it is not optional.
In general, if we have a non-optional result, we always either have to fail within the parser, or we have to correctly read the result's Type. Without the `resultOptional` argument enforcing the reading of the result, we can have ops slip through the cracks by using the form without parens and result Types.

Reply:
I see. In that case there is no way for the parser to know that the result is required. Considering that these ops must have a result, perhaps it would be better for these transform ops to use some different type of assembly format (that requires a result) instead of SemiFunctionType, but the extra option seems okay to me.